pkg/apply/mutator/apply_time_mutator.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package mutator
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cli-utils/pkg/apply/cache"
+	"sigs.k8s.io/cli-utils/pkg/kyq"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/object/mutation"
+)
+
+// ApplyTimeMutator mutates a resource by injecting values specified by the
+// apply-time-mutation annotation.
+// The optional ResourceCache will be used to speed up source resource lookups,
+// if specified.
+// Implements the Mutator interface
+type ApplyTimeMutator struct {
+	Client        dynamic.Interface
+	Mapper        meta.RESTMapper
+	ResourceCache cache.ResourceCache
+}
+
+// Name returns a mutator identifier for logging.
+func (atm *ApplyTimeMutator) Name() string {
+	return "ApplyTimeMutator"
+}
+
+// Mutate parses the apply-time-mutation annotation and loops through the
+// substitutions, applying each of them to the supplied target object.
+// Returns true with a reason, if mutation was performed.
+func (atm *ApplyTimeMutator) Mutate(obj *unstructured.Unstructured) (bool, string, error) {
+	mutated := false
+	reason := ""
+
+	if !mutation.HasAnnotation(obj) {
+		return mutated, reason, nil
+	}
+
+	targetRef := mutation.NewResourceReference(obj)
+
+	subs, err := mutation.ReadAnnotation(obj)
+	if err != nil {
+		return mutated, reason, fmt.Errorf("failed to read jsonpath field in target resource (%v): %w", targetRef, err)
+	}
+
+	targetID := object.UnstructuredToObjMetaOrDie(obj)
+	klog.V(5).Infof("target resource %s/%s: %#v", targetID.Namespace, targetID.Name, obj)
+
+	for _, sub := range subs {
+		sourceRef := sub.SourceRef
+
+		// lookup source resource from cache or cluster
+		sourceObj, err := atm.getObject(sourceRef)
```

Comment:
Maybe I missed it, but I don't see a check for object status in getObject or here, is this dependent fully on the dependsOn implementation to assert readiness prior to fetching? will that make it more challenging in the future to move to supporting mutation references to objects outside the apply set?

Reply:
The depends-on impl added a `graph.SortObjs` that handles ordering of apply stages with wait tasks in between. I extended the SortObjs impl to inject the source resources from apply-time-mutation into the graph:
https://github.com/karlkfi/cli-utils/blob/karl-apply-time-mutation/pkg/object/graph/sort.go#L33

So by the time the mutator is executed in the apply task, the waits should have already happened.

There might be some edge cases with resources not in the inventory. So I'll need to write some tests that exercise both pieces together. But I'm not sure where to put them yet in cli-utils.